File: src/Runtime/python/PyOMCompile.py
#!/usr/bin/env python3

############# PyOMCompile.py #####################################################
#
# Copyright 2026 The IBM Research Authors.
#
################################################################################
# Common class `PyOMCompile` called by python scripts
#
# Use this indirection to call PyOMCompileC in case we need to add some
# specific python processing. None is needed at this time.
################################################################################
import numpy as np

try:
    from PyOMCompileC import OMCompile as OMCompile_

except ImportError:
    raise ImportError(
        "Looks like you did not build the PyOMCompileC target, build it by\n"
        "running `make PyOMCompileC`. You may need to set PYTHONPATH to\n"
        "include `onnx-mlir/build/Debug/lib` since `make PyOMCompileC` outputs\n"
        "to `build/Debug` by default (or Release if building Release)."
    )


class OMCompile(OMCompile_):
    def __init__(
        self,
        compiler_path="",
        compiler_image="",
        auto_pull=True,
        engine="auto",
        cache=None,
        verbose=False,
    ):

        self.cache = cache
        if compiler_image:
            super().__init__(compiler_image, compiler_path, engine, auto_pull, verbose)
        else:
            if not compiler_path:
                # Import the package for standalone compile
                try:
                    import OMPyCompile

                    compiler_path = OMPyCompile.get_compiler_path()
                except ImportError as e:
                    # No standalone compiler, use local PATH
                    pass
            super().__init__(compiler_path, verbose)

    def compile(
        self,
        model_path,
        flags,
        output_path="",
        compiler_path="",
        log_file_name="",
        reuse_compiled_model=False,
    ):
        if self.cache:
            # Check cache policy to decide whether a cached .so can be used
            exit(-1)
        return super().compile(
            model_path,
            flags,
            output_path,
            compiler_path,
            log_file_name,
            reuse_compiled_model,
        )

File: src/Runtime/python/PyOMCompile.cpp
/*
 * SPDX-License-Identifier: Apache-2.0
 */

//===---- PyOMCompile.cpp - PyOMCompile Implementation ------===//
//
// Copyright 2024-2026 The IBM Research Authors.
//
// =============================================================================
//
// This file contains implementations of PyOMCompile class,
// which helps python programs to compile and run binary model libraries.
//
//===----------------------------------------------------------------------===//

#include "PyOMCompile.hpp"

namespace onnx_mlir {

// =============================================================================
// Constructors

// Constructor for local compilation
PyOMCompile::PyOMCompile(const std::string &compilerPath, bool verbose)
    : OMcompile(compilerPath, verbose) {}

// Constructor for container-based compilation
PyOMCompile::PyOMCompile(const std::string &containerImage,
    const std::string &compilerPathInContainer, const std::string &engine,
    bool autoPull, bool verbose)
    : OMcompile(containerImage, compilerPathInContainer,
          engine == "docker"   ? OMCompile::ContainerEngine::Docker
          : engine == "podman" ? OMCompile::ContainerEngine::Podman
                               : OMCompile::ContainerEngine::Auto,
          autoPull, verbose) {}

// =============================================================================
// Compile method

std::string PyOMCompile::compile(const std::string &modelPath,
    const std::string &flags, const std::string &outputPath,
    const std::string &compilerPath, const std::string &logFilename,
    bool reuseCompiledModel) {
  std::string filename = pyPredictOutputFilename(modelPath, flags, outputPath);
  if (reuseCompiledModel) {
    if (!filename.empty()) {
      FILE *file = fopen(filename.c_str(), "r");
      if (file) {
        // File exists, no compilation is needed
        fclose(file);
        return filename;
      }
    }
  }

  // Let compilation exceptions propagate naturally to Python without
  // printing to stderr. Python code can handle and display exceptions
  // as needed, avoiding duplicate error messages.
  // Old version caught and re-threw with stderr output, causing duplicates.
  OMcompile.compile(modelPath, flags, outputPath, compilerPath, logFilename);
  // Check for compiler consistency
  assert(filename == OMcompile.getOutputFilename() &&
         "Something wrong with OMCompile.cpp");
  return OMcompile.getOutputFilename();
}

// =============================================================================
// Getters

std::string PyOMCompile::pyGetOutputFilename() {
  return OMcompile.getOutputFilename();
}

/* static */ std::string PyOMCompile::pyPredictOutputFilename(
    const std::string &modelPath, const std::string &flags,
    const std::string &outputPath) {
  return onnx_mlir::OMCompile::predictOutputFilename(
      modelPath, flags, outputPath);
}

std::string PyOMCompile::pyGetOutputConstantFilename() {
  return OMcompile.getOutputConstantFilename();
}

std::string PyOMCompile::pyGetModelTag() { return OMcompile.getModelTag(); }

bool PyOMCompile::pyIsSuccessfullyCompiled() {
  return OMcompile.isSuccessfullyCompiled();
}

bool PyOMCompile::pyHasOutputConstantFilename() {
  return OMcompile.hasOutputConstantFilename();
}

} // namespace onnx_mlir
